File: benchmarks/fixtures/webapp_ts/src/tasks/cleanupTask.ts
/** Cleanup background tasks. */
import { getLogger } from '../utils/helpers';
import { DatabaseConnection } from '../database/connection';
import { CacheBackend } from '../cache/base';

const logger = getLogger("tasks.cleanup");

/** Clean up expired sessions. */
export async function cleanupExpiredSessions(db: DatabaseConnection): Promise<number> {
    logger.info("Cleaning up expired sessions");
    const result = await db.executeQuery("UPDATE sessions SET expiredAt = ? WHERE expiredAt IS NULL AND createdAt < ?", [Date.now(), Date.now() - 7 * 86400 * 1000]);
    logger.info(`Expired ${result.affected} sessions`);
    return result.affected;
}

/** Clean up old events. */
export async function cleanupOldEvents(db: DatabaseConnection): Promise<number> {
    logger.info("Cleaning up old events");
    const result = await db.executeQuery("DELETE FROM events WHERE processedAt IS NOT NULL AND createdAt < ?", [Date.now() - 30 * 86400 * 1000]);
    return result.affected;
}

/** Flush cache. */
export function cleanupCache(cache: CacheBackend): number {
    logger.info("Running cache cleanup");
    return cache.clear();
}

/** Run all cleanup tasks. */
export async function runAllCleanup(db: DatabaseConnection, cache: CacheBackend): Promise<Record<string, number>> {
    const sessions = await cleanupExpiredSessions(db);
    const events = await cleanupOldEvents(db);
    const cacheEntries = cleanupCache(cache);
    logger.info("Cleanup complete");
    return { expiredSessions: sessions, oldEvents: events, cacheCleared: cacheEntries };
}

File: benchmarks/fixtures/webapp_rb/tasks/cleanup_task.rb
# Cleanup background tasks.

require_relative '../utils/helpers'
require_relative '../database/connection'
require_relative '../cache/base_cache'

# Clean up expired sessions.
def cleanup_expired_sessions(db)
  get_logger('tasks.cleanup').info('Cleaning up expired sessions')
  result = db.execute_query(
    'UPDATE sessions SET expired_at = ? WHERE expired_at IS NULL AND created_at < ?',
    [Time.now.to_i, Time.now.to_i - 7 * 86_400]
  )
  get_logger('tasks.cleanup').info("Expired #{result[:affected]} sessions")
  result[:affected]
end

# Clean up old events.
def cleanup_old_events(db)
  get_logger('tasks.cleanup').info('Cleaning up old events')
  result = db.execute_query(
    'DELETE FROM events WHERE processed_at IS NOT NULL AND created_at < ?',
    [Time.now.to_i - 30 * 86_400]
  )
  result[:affected]
end

# Flush cache.
def cleanup_cache(cache)
  get_logger('tasks.cleanup').info('Running cache cleanup')
  cache.clear
end

# Run all cleanup tasks.
def run_all_cleanup(db, cache)
  sessions = cleanup_expired_sessions(db)
  events = cleanup_old_events(db)
  cache_entries = cleanup_cache(cache)
  get_logger('tasks.cleanup').info('Cleanup complete')
  { expired_sessions: sessions, old_events: events, cache_cleared: cache_entries }
end
